### PR TITLE
Remove mask slicing in all eager attentions

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title. Following https://github.com/huggingface/transformers/pull/41900.

The mask is (and should!!) be correctly prepared, with the correct shape. If not, then it0s better to crash immediately, as otherwise this leads to very silent bugs!!

### CODE REVIEW

**Comment by justinchuby** [comment]
Was this able to move forward? Thanks!

**Comment by Cyrilvallez** [comment]
Hey @justinchuby! This has been on standby a bit as a lot of errors popped up, because some old models do not create their masks correctly... 
We are right now removing old mask preparation primitives from every model, so once this is done this PR should be able to land immediately!

**Comment by Cyrilvallez** [comment]
Finally ready now that https://github.com/huggingface/transformers/pull/42848 was merged

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -274,7 +274,7 @@ def forward(
 
         # change view to [batch_size, num_heads, q_length, kv_length]
         attn_weights = attention_scores.view(batch_size, self.num_heads, q_length, -1)
-        if attention_mask is not None:  # no matter the length, we just slice it
+        if attention_mask is not None:
             causal_mask = attention_mask[:, :, :, : key_layer.shape[-1]]
```
Comment (vasqu) [inline src/transformers/models/bloom/modeling_bloom.py]: Oops :D

File: src/transformers/models/diffllama/modeling_diffllama.py [hunk 2]
```diff
@@ -255,7 +255,7 @@ def forward(
 
         attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
 
-        if attention_mask is not None:  # no matter the length, we just slice it
+        if attention_mask is not None:
             causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
```
Comment (vasqu) [inline src/transformers/models/diffllama/modeling_diffllama.py]: Same

File: src/transformers/models/diffllama/modular_diffllama.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def forward(
 
         attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
 
-        if attention_mask is not None:  # no matter the length, we just slice it
+        if attention_mask is not None:
             causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]
```
Comment (vasqu) [inline src/transformers/models/diffllama/modular_diffllama.py]: Ok mentioning it a last time lol

**Review by vasqu — COMMENTED** [review]
Wdyt about using 2 sources (because some do fp32 softmax and some don't) for eager to unify this? Talking about those that are not using modular atm --> use copied from to ensure we only modify these

Just a thought which might make maintenance easier + I don't have a good overview on which are independent atm

File: src/transformers/models/bloom/modeling_bloom.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/bloom/modeling_bloom.py]: Damn nice catch! This one thought it could keep on living... 😈

**Comment by Cyrilvallez** [comment]
I agree that in general it would be very nice to unify a bit more the eager implementations which are all the same but use different variable names etc for the same things.
This feels a bit outside the scope of this PR which touches a LOT of files already though, and which is focused on removing a useless and dangerous slicing ops - would prefer to do it later on to avoid overcomplicating this one

**Review by vasqu — APPROVED** [review]
Yup, it's fine that way - wanna run a few slow tests (run-slow) on important models for safety? Otherwise lgtm

**Comment by Cyrilvallez** [comment]
run-slow: bert, gemma2, llama, mistral, mixtral
